source/dub/project.d

```diff
@@ -643,64 +645,91 @@ class Project {
-		string[][string] parents;
-		parents[m_rootPackage.name] = null;
-		foreach (p; getTopologicalPackageList())
-			foreach (d; p.getAllDependencies())
-				parents[d.name.toString()] ~= p.name;
 
-		size_t createConfig(string pack, string config) {
+		// cached package information to avoid continuous re-computation
+		// during the resolution process
+		size_t[string] package_map;
+		size_t[][] parents;
+		const(Package)[] package_list;
+		string[] package_names;
+		PackageDependency[][] package_dependencies;
+		package_list.reserve(m_dependencies.length);
+		package_names.reserve(m_dependencies.length);
+		package_dependencies.reserve(m_dependencies.length);
+		foreach (p; getTopologicalPackageList()) {
+			auto pname = p.name;
+			package_map[pname] = package_list.length;
+			package_list ~= p;
+			package_names ~= pname;
+			package_dependencies ~= p.getAllDependencies();
+		}
+		parents.length = package_list.length;
+		foreach (pack_idx, pack_deps; package_dependencies) {
+			foreach (d; pack_deps)
+				if (auto pi = d.name.toString() in package_map)
+					parents[*pi] ~= pack_idx;
+		}
+
+		size_t createConfig(size_t pack_idx, string config) {
 			foreach (i, v; configs)
-				if (v.pack == pack && v.config == config)
+				if (v.pack == pack_idx && v.config == config)
 					return i;
-			assert(pack !in m_overriddenConfigs || config == m_overriddenConfigs[pack]);
-			logDebug("Add config %s %s", pack, config);
-			configs ~= Vertex(pack, config);
-			return configs.length-1;
-		}
 
-		bool haveConfig(string pack, string config) {
-			return configs.any!(c => c.pack == pack && c.config == config);
+			auto pname = package_names[pack_idx];
+			assert(pname !in m_overriddenConfigs || config == m_overriddenConfigs[pname]);
+			logDebug("Add config %s %s", pname, config);
+			auto cfg = Vertex(pack_idx, config);
+			configs ~= cfg;
+			configs_set[cfg] = (void[0]).init;
+			return configs.length-1;
 		}
 
-		size_t createEdge(size_t from, size_t to) {
-			auto idx = edges.countUntil(Edge(from, to));
-			if (idx >= 0) return idx;
-			logDebug("Including %s %s -> %s %s", configs[from].pack, configs[from].config, configs[to].pack, configs[to].config);
-			edges ~= Edge(from, to);
-			return edges.length-1;
+		bool haveConfig(size_t pack_idx, string config) {
+			return (Vertex(pack_idx, config) in configs_set) !is null;
 		}
 
 		void removeConfig(size_t i) {
 			logDebug("Eliminating config %s for %s", configs[i].config, configs[i].pack);
 			auto had_dep_to_pack = new bool[configs.length];
 			auto still_has_dep_to_pack = new bool[configs.length];
 
-			edges = edges.filter!((e) {
-					if (e.to == i) {
-						had_dep_to_pack[e.from] = true;
-						return false;
-					} else if (configs[e.to].pack == configs[i].pack) {
-						still_has_dep_to_pack[e.from] = true;
-					}
-					if (e.from == i) return false;
-					return true;
-				}).array;
+			// eliminate all edges that connect to config 'i'
+			size_t eti = 0;
+			foreach (ei, e; edges) {
```

Comment:
I don't think this for loop is small enough to justify variable names such as `eti`, `ei`, and `e`.

Reply:
The loop contains 8 lines of code. What's your personal threshold?

Reply:
> The loop contains 8 lines of code. What's your personal threshold?

For one letter variables? 1 line.

Reply:
For me, a handful of lines is okay, as long variable names are not only making the code easier to understand, but also harder to read - although non-standard variables like the ones here are more critical than the standard ones, if used correctly (i, j, k , x, y, z). However, the parameter `i` to `removeConfig` actually violates my own style guide and should be a full word instead, which would free up the `i` for the `foreach` loop.

But IMO, what would really be the right way to go here would be to generalize the array filter algorithm, so that this becomes `edges = edges.filterInPlace!((e) => ...)` (analogous to `std.algorithm.mutation.remove`). Unfortunately, Phobos doesn't have that AFAIK and I didn't want to mix non-local refactoring (or any substantial refactoring for that matter) into the same PR that does optimization.